waspc/data/Generator/templates/react-app/src/test/vitest/helpers.tsx

```diff
@@ -0,0 +1,72 @@
+import React from 'react'
+import { rest } from 'msw'
+import { setupServer } from 'msw/node'
+import { BrowserRouter as Router } from 'react-router-dom'
+import { render, RenderResult, cleanup } from '@testing-library/react'
+import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
+import { beforeAll, afterEach, afterAll } from 'vitest'
+import { Query } from '../../queries'
+import config from '../../config'
+import { type HttpMethod } from '../../types'
+
+export function renderInContext(ui: React.ReactElement): RenderResult {
+  const client = new QueryClient()
+  const { rerender, ...result } = render(
+    <QueryClientProvider client={client}><Router>{ui}</Router></QueryClientProvider>
+  )
+  return {
+    ...result,
+    rerender: (rerenderUi: React.ReactElement) =>
+      rerender(
+        <QueryClientProvider client={client}><Router>{rerenderUi}</Router></QueryClientProvider>
+      )
+  }
+}
+
+type QueryRoute = Query<any, any>['route']
+type Handler = () => void
+
+export function mockServer() {
+  const server = setupServer()
+
+  beforeAll(() => server.listen())
+  afterEach(() => {
+    server.resetHandlers()
+    cleanup()
+  })
+  afterAll(() => server.close())
+
+  function mockQuery({ route }: { route: QueryRoute }, resJson: any): void {
+    const url = `${config.apiUrl}${route.path}`
+    const responseHandler = (_req, res, ctx) => {
+      return res(ctx.json(resJson))
+    }
+
+    // NOTE: Technically, we only need to care about POST for Queries
+    // and GET for the /auth/me route. However, an additional use case
+    // for this function could be to mock APIs, so more methods are supported.
+    const handlers: Record<HttpMethod, Handler> = {
+      GET: () => {
+        server.use(rest.get(url, responseHandler))
+      },
+      POST: () => {
+        server.use(rest.post(url, responseHandler))
+      },
+      PUT: () => {
+        server.use(rest.put(url, responseHandler))
+      },
+      DELETE: () => {
+        server.use(rest.delete(url, responseHandler))
+      },
+    }
+
+    const setupMock = handlers[route.method]
+    if (!setupMock) {
+      throw new Error(`Unsupported query method for mocking: ${route.method}. Supported method strings are: ${Object.keys(handlers).join(', ')}.`)
+    }
+
+    setupMock()
```

Comment:
You could simplify this block to:
```typescript
const handlers: Record<HttpMethod, Parameters<typeof server.use>[0]> = {
  GET: rest.get(url, responseHandler),
  POST: rest.post(url, responseHandler),
  PUT: rest.put(url, responseHandler),
  DELETE: rest.delete(url, responseHandler),
}

server.use(handlers[route.method])
```

## What about checking for the wrong method?
I've removed the part where you check for whether the correct method was specified because TypeScript takes care of it.

I assume you wanted to make sure to return an informative message even when the caller isn't using TypeScript. If that's the case, it's definitely better to keep the check.  However, we should perform it directly and as soon as possible (instead of relying on a map lookup to fail).

To perform this check without having to define a handler mapping first, you'll need a way to access the allowed methods in **the value space** (the type `HttpMethod` only introduces them to the type space).

This is where enums really shine (we've already used them [here](https://github.com/wasp-lang/wasp/pull/961#discussion_r1086804020)):
```typescript
enum HttpMethod {
    Get = 'GET',
    Post = 'POST',
    Put = 'PUT',
    Delete = 'DELETE',
}

// ...

function mockQuery({ route }: { route: QueryRoute }, resJson: any): void {
  if(!Object.values(HttpMethod).includes(route.method)) {
    // throw error
  }

  // continue knowing that route.method is a real method
  // ...

  const handlers: Record<HttpMethod, Parameters<typeof server.use>[0]> = {
    [HttpMethod.Get]: rest.get(url, responseHandler),
    [HttpMethod.Post]: rest.post(url, responseHandler),
    [HttpMethod.Put]: rest.put(url, responseHandler),
    // You can still use the string directly if you like
    // but the above method is easier to auto-refactor
    DELETE: rest.delete(url, responseHandler),
  }

  server.use(handlers[route.method])
}
```
> **Note**
> Enums in TypeScript are pretty weird. [This StackOverflow answer](https://stackoverflow.com/a/50396312) explains them nicely. If you didn't need runtime checks, a union type (like the one Miho recommended and you have now) would be a better choice.

Reply:
Great tips here @sodic, thanks, and using it in deploy is definitely coming back to me! :D Eventually it will become natural haha 🙏🏻